core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1556,6 +1572,8 @@ class Log(@volatile private var _dir: File,
           done = fetchDataInfo != null || segmentEntry == null
         }
 
+        checkForLogDirFailure()
```

Comment:
Seems more intuitive to move this check before the segment read. I don't think we can totally avoid race conditions with a failure in append since we don't have the lock here. Perhaps we could even move this check to `maybeHandleIOException` so that we handle all cases?

Reply:
Yeah, I wondered about putting it in `maybeHandleIOException` too, and I've made this change. There are a few places where changing the order in which we do `maybeHandleIOException` with respect to `lock synchronized` would seem to avoid some possible races with no side effects other than increasing the duration the lock is held (for the `try` and the volatile read). I'm thinking specifically `maybeIncrementLogStartOffset`, `roll`, `delete`, `truncateFullyAndStartAt` and perhaps even `truncateTo`. But given that there will always exist some races maybe it's not worth it, WDYT?